Get measurements

1r-1na opened the pull request:
## Check hypertables
- with command `select * from _timescaledb_catalog.hypertable` in a tenant db it is possiblee to check for hypertables.
![image](https://github.com/user-attachments/assets/5bbb132f-da0f-4328-963d-47287ce696b7)

1r-1na committed:
feat(database): Change db image to timescale and all dependent changes

1r-1na committed:
feat(database): Added method to ensure hypertable

1r-1na committed:
feat(database): added index for column timestamp in table measurement

1r-1na committed:
Merge remote-tracking branch 'origin/main' into feature/81-backend-get-measurements

1r-1na committed:
feat(database): added migration for index on column timestamp in measurement

1r-1na committed:
feat(measurement): removed `MeasurementData` and added the properties to `Measurement`

1r-1na committed:
feat(measurement): added migration to declare table `Measurement` as hypertable

1r-1na committed:
Merge remote-tracking branch 'origin/main' into feature/80-backend-get-measurements

1r-1na committed:
feat(measurement): added index to smart meter in measurement table and query for measurements

1r-1na committed:
feat(measurement): added endpoint for measurements

Smightym8 commented on src/SMAIAXBackend.API/Endpoints/Measurement/GetMeasurementsEndpoint.cs:
```diff
@@ -0,0 +1,21 @@
+        [FromQuery] DateTime? startAt, [FromQuery] DateTime? endAt)
+    {
+        var measurements =
+            await measurementListService.GetFilteredMeasurementsByTenantAndSmartMeterAsync(
```
Rename `GetFilteredMeasurementsByTenantAndSmartMeterAsync` to `GetFilteredMeasurementsBySmartMeterAsync`.

Smightym8 commented on src/SMAIAXBackend.API/Middlewares/ExceptionHandlerMiddleware.cs:
```diff
@@ -42,6 +42,7 @@ public async ValueTask<bool> TryHandleAsync(
                 problemDetails.Status = StatusCodes.Status409Conflict;
                 problemDetails.Title = "Conflict";
                 break;
+            case ArgumentException:
```
Remove the `ArgumentException` and use only custom exceptions here.

Smightym8 commented on src/SMAIAXBackend.Application/Services/Implementations/MeasurementListService.cs:
```diff
@@ -0,0 +1,29 @@
+    IMeasurementRepository measurementRepository,
+    ISmartMeterRepository smartMeterRepository) : IMeasurementListService
+{
+    public async Task<List<MeasurementRawDto>> GetFilteredMeasurementsByTenantAndSmartMeterAsync(
```
Rename `GetFilteredMeasurementsByTenantAndSmartMeterAsync` to `GetFilteredMeasurementsBySmartMeterAsync`.

Smightym8 commented on src/SMAIAXBackend.Application/Services/Interfaces/IMeasurementListService.cs:
```diff
@@ -0,0 +1,19 @@
+    /// </param>
+    /// <param name="endAt">Optional timestamp filter. Data with a timestamp older/smaller than endAt will be returned.</param>
+    /// <returns>All measurements between given limitations or by default measurements of one day.</returns>
+    Task<List<MeasurementRawDto>> GetFilteredMeasurementsByTenantAndSmartMeterAsync(Guid smartMeterId,
```
Rename `GetFilteredMeasurementsByTenantAndSmartMeterAsync` to `GetFilteredMeasurementsBySmartMeterAsync`.

Smightym8 commented on src/SMAIAXBackend.Domain/Repositories/IMeasurementRepository.cs:
```diff
@@ -0,0 +1,20 @@
+    /// </param>
+    /// <param name="endAt">Optional timestamp filter. Data with a timestamp older/smaller than endAt will be returned.</param>
+    /// <returns>All measurements between given limitations or by default measurements of one day.</returns>
+    Task<List<Measurement>> GetMeasurementsByTenantAndSmartMeterAsync(SmartMeterId smartMeterId,
```
Rename `GetMeasurementsByTenantAndSmartMeterAsync` to `GetMeasurementsBySmartMeterAsync`.

Smightym8 commented on src/SMAIAXBackend.Infrastructure/Repositories/MeasurementRepository.cs:
```diff
@@ -0,0 +1,34 @@
+        DateTime? startAt = null,
+        DateTime? endAt = null)
+    {
+        if (endAt > startAt)
```
Move this check to the application service and throw a custom exception.

Smightym8 commented on src/SMAIAXBackend.Infrastructure/Repositories/MeasurementRepository.cs:
```diff
@@ -0,0 +1,34 @@
+        }
+
+        // If endAt is null, by default one month is returned, either from startAt or from DateTime.UtcNow.
+        endAt ??= startAt?.AddDays(-1) ?? DateTime.UtcNow.AddDays(-1);
```
Move this to the application service.

1r-1na replied:
I disabled the automatic range to one day and made startAt and endAt required.

1r-1na committed:
feat(measurement): Resolved comments and added tests
- Renamed `GetFilteredMeasurementsByTenantAndSmartMeterAsync` to `GetMeasurementsBySmartMeterAsync`
- Made startAt and endAt query parameters required
- Added integration tests for MeasurementRepository and GetMeasurementsEndpoint.
- Introduced InvalidTimeRangeException

Smightym8 committed:
refactor(measurement): address SonarCloud warnings and improve exception handling

- Renamed test cases to align with the existing naming pattern.
- Moved the static message for `InvalidTimeRangeException` into the exception class, as the message remains constant for all occurrences of this exception.

Smightym8 committed:
style: format code

Smightym8 committed:
fix(tests): fix failing tests

- Fixed wrong sql queries in TenantDbContext and TestBase
- Adjusted date of measurement and in tests so it is always within the queried time range

Smightym8 committed:
test(measurement): Add unit tests for MeasurementListService

1r-1na committed:
feat(measurement): added MeasurementRepositoryTests when no measurements in time range

1r-1na committed:
resolved(measurement): resolved conflicts
